src/PowerShellEditorServices/Session/Host/TerminalPSHostRawUserInterface.cs

```diff
@@ -184,11 +145,87 @@ public override Size MaxWindowSize
+                ControlKeyStates states = default;
+                if ((key.Modifiers & ConsoleModifiers.Alt) != 0)
+                {
+                    states |= ControlKeyStates.LeftAltPressed;
+                }
+
+                if ((key.Modifiers & ConsoleModifiers.Control) != 0)
+                {
+                    states |= ControlKeyStates.LeftCtrlPressed;
+                }
+
+                if ((key.Modifiers & ConsoleModifiers.Shift) != 0)
+                {
+                    states |= ControlKeyStates.ShiftPressed;
+                }
 
-            throw new System.NotImplementedException();
+                var result = new KeyInfo((int)key.Key, key.KeyChar, states, isDown);
+                if (isDown)
+                {
+                    this.lastKeyDown = result;
+                }
+
+                return result;
+            }
+
+            bool includeUp = (options & ReadKeyOptions.IncludeKeyUp) != 0;
+
+            // Key Up was requested and we have a cached key down we can return.
+            if (includeUp && this.lastKeyDown != null)
+            {
+                KeyInfo info = this.lastKeyDown.Value;
+                this.lastKeyDown = null;
+                return new KeyInfo(
+                    info.VirtualKeyCode,
+                    info.Character,
+                    info.ControlKeyState,
+                    keyDown: false);
+            }
+
+            bool intercept = (options & ReadKeyOptions.NoEcho) != 0;
+            bool includeDown = (options & ReadKeyOptions.IncludeKeyDown) != 0;
+            if (!(includeDown || includeUp))
+            {
+                throw new PSArgumentException(
+                    "Cannot read key options. To read options, set one or both of the following: IncludeKeyDown, IncludeKeyUp.",
+                    nameof(options));
+            }
+
+            // Allow ControlC as input so we can emulate pipeline stop requests. We can't actually
+            // determine if a stop is requested without using non-public API's.
+            bool oldValue = System.Console.TreatControlCAsInput;
+            try
+            {
+                System.Console.TreatControlCAsInput = true;
+                ConsoleKeyInfo key = ConsoleProxy.ReadKey(intercept, default(CancellationToken));
+
+                if (IsCtrlC(key))
+                {
+                    // Caller wants CtrlC as input so return it.
+                    if ((options & ReadKeyOptions.AllowCtrlC) == 0)
+                    {
+                        return ProcessKey(key, includeDown);
+                    }
+
+                    // Caller doesn't want CtrlC so throw a PipelineStoppedException to emulate
+                    // a real stop.  This will not show an exception to a script based caller and it
+                    // will avoid having to return something like default(KeyInfo).
+                    throw new PipelineStoppedException();
```

Comment:
> Caller doesn't want CtrlC so throw a PipelineStoppedException to emulate a real stop.

@SeeminglyScience I'm a little confused about the comment on this throw statement. Can you please explain a bit?

Reply:
Sure, so at that point we've received the input <kbd>Ctrl</kbd> + <kbd>C</kbd>.  If the caller did not include `AllowCtrlC` in the options, then we should treat it like a break event instead and terminate the pipeline.

Emulate is probably not *really* the right word there, I chose it because we aren't going through `PowerShell.Stop` or `Pipeline.Stop` like a normal break event would.

Another route would be to not set `Console.TreatControlCAsInput`, and allow our break handler to call `PowerShell.Stop` like normal.  This actually would be more similar to what `ConsoleHost` does as it will actually return `default(KeyInfo)`, but triggers a pipeline stop.  The issue I have with this is that if `RawUI.ReadKey` is in an assignment statement, the pipeline stop won't actually happen until the *next* sequence point.  PowerShell doesn't check if the current command is stopping between the RHS being evaluated and it's result being set to the LHS.  So the LHS will actually be set to `default(KeyInfo)` instead of remaining unchanged.  That doesn't happen if we throw instead.

Reply:
> Another route would be to not set Console.TreatControlCAsInput

Thanks! My confusion was mainly around this case. It makes sense now.

> So the LHS will actually be set to default(KeyInfo) instead of remaining unchanged.

Why does this matter if the pipeline stop will kick in at the next sequence point? Even if it's set to `default(KeyInfo)`, the script execution gets killed and that assignment shouldn't have any real effect anymore, right?

Reply:
> the script execution gets killed and that assignment shouldn't have any real effect anymore, right?

There's a few scenarios where it can linger:

1. LHS is a global scoped variable (or script scoped variable in a module)
1. LHS is a property or index expression
1. Current command is a prompt evaluation (or is otherwise running within the top level scope of it's `SessionState`)

Reply:
Here's the easiest way:

```powershell
$key = $null
# Drain "Enter" key up from accepting prompt.  Maybe comment
# this line for non-Windows.
$null = $Host.UI.RawUI.ReadKey('NoEcho, IncludeKeyUp')
$key = $Host.UI.RawUI.ReadKey('NoEcho, IncludeKeyUp')
```

Then `$key` will hold

```raw
VirtualKeyCode Character ControlKeyState KeyDown
-------------- --------- --------------- -------
             0                         0   False
```

Reply:
OK, I can repro it now after removing `PSReadline`.
Interesting that if I change `IncludeKeyUp` to `IncludeKeyDown`, it doesn't repro anymore
```
$null = $Host.UI.RawUI.ReadKey('NoEcho, IncludeKeyDown')
$key = $Host.UI.RawUI.ReadKey('NoEcho, IncludeKeyDown')

VirtualKeyCode Character            ControlKeyState KeyDown
-------------- ---------            --------------- -------
            17           LeftCtrlPressed, NumLockOn    True
```

Reply:
BTW, my comment is not blocking on this PR. I'm just curious ...

Reply:
> OK, I can repro it now after removing PSReadline.

Huh that's surprising... mine was with PSRL as well.  Though I'm on 2.0 beta 3 (or more likely one of the appveyor builds...)

> Interesting that if I change IncludeKeyUp to IncludeKeyDown, it doesn't repro anymore

Wellll sorta.  `IncludeKeyDown` makes it a lot harder to reproduce because it can trigger on a single press of <kbd>Ctrl</kbd>.  But if you do something like this, it'll still show up

```powershell
$key = $null
while ($true) {
    $key = $Host.UI.RawUI.ReadKey('NoEcho, IncludeKeyDown')
}
```

That will continously `ReadKey` until it registers <kbd>Ctrl</kbd> + <kbd>C</kbd> and a pipeline stop is triggered.

> BTW, my comment is not blocking on this PR. I'm just curious :)

No worries, I could talk about it forever 🙂

Reply:
> Huh that's surprising... mine was with PSRL as well. Though I'm on 2.0 beta 3 (or more likely one of the appveyor builds...)

Ha, I'm using `6.2.0-rc.1`.

> IncludeKeyDown makes it a lot harder to reproduce because it can trigger on a single press of Ctrl. So by the time you type C, you're actually already back to a prompt.

That makes perfect sense!